### PR TITLE
BUG: freeze setuptools build deps to workaround pypa/setuptools#2849

### PR DESCRIPTION
**Pull request by skirpichev**
xref pypa/setuptools#2849

As seen in [recent](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=14690&view=logs&jobId=ede89041-5e8e-5201-ce5c-f68c2cbf02dd&j=ede89041-5e8e-5201-ce5c-f68c2cbf02dd&t=d29e43e0-51b0-592c-77cf-c2635d71df89) CI jobs, dependency installation is failing with
```
    AttributeError: get_data_files_without_manifest
```

### CODE REVIEW

File: azure-pipelines.yml
Comment (tupui) [inline azure-pipelines.yml:188]: Any reason this is different from the other?
```suggestion
            pip3 install setuptools<58.5 wheel --user && \
```

**Review by tupui — COMMENTED** [review]
Thanks @skirpichev. Indeed recent CI jobs are failing with this error.

File: azure-pipelines.yml
Comment (skirpichev) [reply azure-pipelines.yml:188]: I think this could fix failures in azure-pipelines.   Lets see.

Comment (skirpichev) [reply azure-pipelines.yml:188]: BTW, "pip3 install setuptools<58.5 wheel --user &&" should do same (I'll use this in the final version).

The tricky part is why the pip deps resolver substitute setuptools>=58.5 (probably due to indirect deps in the matplotlib) even if we have strict requirement for setuptools==58.4.0.  That's why we can't install all things in one shot.  Seems to be a pip bug...

**Comment by skirpichev** [comment]
Ok, now all tests - pass.  I did rebase to clean up WIP commits.

**Comment by tupui** [comment]
> Ok, now all tests - pass. I did rebase to clean up WIP commits.

I will merge once green to unlock others. We should keep an issue open though to not forget this.

**Comment by skirpichev** [comment]
@tupui, there is no Scipy issue so far.

**Comment by tupui** [comment]
> @tupui, there is no Scipy issue so far.

I know, you can create one if you want 😃 Otherwise I will do this after I merge.

**Comment by skirpichev** [comment]
Honestly, I'm not sure how it should be.  IMHO, freezing deps like this is (1) a good idea in general and (2) a self-documenting thing :)

**Comment by tupui** [comment]
> Honestly, I'm not sure how it should be. IMHO, freezing deps like this is (1) a good idea in general and (2) a self-documenting thing :-)

We don't want to freeze non direct dependencies like that. It brings too much issues for downstream packages if we do.

**Comment by rgommers** [comment]
> Honestly, I'm not sure how it should be. IMHO, freezing deps like this is (1) a good idea in general and (2) a self-documenting thing :-)

We do put an upper bound on dependencies in release branches. In master we want to catch problems early, so they can be fixed.

Agreed, doesn't need separate docs nor a separate issue. Let's just revert it as soon as it's fixed, which will probably be without one day given the amount of breakage that that release caused.

**Review by rgommers — APPROVED** [review]
LGTM, thanks @skirpichev. And thanks @tupui for reviewing.

**Comment by tupui** [comment]
Wooo everything is actually green?! Did we fix something on our side @rgommers?

**Comment by rgommers** [comment]
Merged, thanks again!

> Wooo everything is actually green?! Did we fix something on our side @rgommers?

The new nightly build of numpy which fixed the custom allocator warnings finally showed up. I'll close the relevant issue.

**Comment by tupui** [comment]
Oh great!

**Comment by skirpichev** [comment]
> Wooo everything is actually green?! 

@tupui  Probably, you miss the green status before rebase.  Unfortunately, setuptools still buggy.

**Comment by tupui** [comment]
> @tupui Probably, you miss the green status before rebase. Unfortunately, setuptools still buggy.

@skirpichev I am talking about something else. We had a job failing for some time related to NumPy as Ralf said.

**Pull request merged**
